File: sdk/python/morego/session.py
from morego.proto import session_pb2
from morego.proto import driver_pb2
from .locator import Locator

class Session:
    def __init__(self, device, session_id, app_id):
        self.device = device
        self.session_id = session_id
        self.app_id = app_id

    def close(self):
        req = session_pb2.CloseSessionRequest(session_id=self.session_id)
        self.device.session_client.CloseSession(req)

    def swipe(self, start_x: int, start_y: int, end_x: int, end_y: int, duration_ms: int):
        req = driver_pb2.SwipeRequest(
            driver_id=self.session_id,
            start=driver_pb2.Point(x=start_x, y=start_y),
            end=driver_pb2.Point(x=end_x, y=end_y),
            duration_ms=duration_ms
        )
        self.device.driver_client.Swipe(req)

    def get_source(self, format_type: str = "xml") -> str:
        req = driver_pb2.GetSourceRequest(
            driver_id=self.session_id,
            format=format_type
        )
        resp = self.device.driver_client.GetSource(req)
        return resp.source_data

    def locator(self, strategy: str, selector: str) -> Locator:
        return Locator(self, strategy, selector)

File: sdk/java/src/main/java/com/morego/sdk/Session.java
package com.morego.sdk;

import github.com.ranjith035.morego.proto.v1.Session.CloseSessionRequest;
import github.com.ranjith035.morego.proto.v1.Driver.SwipeRequest;
import github.com.ranjith035.morego.proto.v1.Driver.GetSourceRequest;
import github.com.ranjith035.morego.proto.v1.Driver.GetSourceResponse;
import github.com.ranjith035.morego.proto.v1.Driver.Point;

public class Session {
    private final Device device;
    private final String sessionId;
    private final String appId;

    public Session(Device device, String sessionId, String appId) {
        this.device = device;
        this.sessionId = sessionId;
        this.appId = appId;
    }

    public String getSessionId() {
        return sessionId;
    }

    public void close() {
        CloseSessionRequest req = CloseSessionRequest.newBuilder()
                .setSessionId(this.sessionId)
                .build();
        this.device.sessionStub.closeSession(req);
    }

    public void swipe(int startX, int startY, int endX, int endY, int durationMs) {
        SwipeRequest req = SwipeRequest.newBuilder()
                .setDriverId(this.sessionId)
                .setStart(Point.newBuilder().setX(startX).setY(startY).build())
                .setEnd(Point.newBuilder().setX(endX).setY(endY).build())
                .setDurationMs(durationMs)
                .build();
        this.device.driverStub.swipe(req);
    }

    public String getSource() {
        return getSource("xml");
    }

    public String getSource(String format) {
        GetSourceRequest req = GetSourceRequest.newBuilder()
                .setDriverId(this.sessionId)
                .setFormat(format)
                .build();
        GetSourceResponse resp = this.device.driverStub.getSource(req);
        return resp.getSourceData();
    }

    public Locator locator(String strategy, String selector) {
        return new Locator(this, strategy, selector);
    }

    protected Device getDevice() {
        return device;
    }
}
